domain/application/state/application_endpoint_test.go

```diff
@@ -0,0 +1,505 @@
+		"unknown": "alpha",
+	}
+
+	// Act
+	err = db.Txn(context.Background(), func(ctx context.Context, tx *sqlair.TX) error {
+		return s.state.insertApplicationEndpoints(context.Background(), tx, insertApplicationEndpointsParams{
+			appID:     coreapplication.ID(s.constants.appID),
+			charmUUID: corecharm.ID(s.constants.charmUUID),
+			bindings:  bindings,
+		})
+	})
+
+	// Assert: should fail because unknown is not a valid relation
+	c.Assert(err, jc.ErrorIs, applicationerrors.CharmRelationNotFound)
+}
+
+// applicationEndpoint represents an association between a charm relation and a
+// specific network space. It is used to fetch the state in order to verify what
+// has been created
+type applicationEndpoint struct {
+	charmRelationUUID string
+	spaceName         string
+}
+
+// fetchApplicationEndpoints retrieves a list of application endpoints from
+// the database based on the application UUID.
+//
+// Returns a slice of applicationEndpoint containing charmRelationUUID and
+// spaceName for each endpoint.
+func (s *applicationEndpointStateSuite) fetchApplicationEndpoints(c *gc.C) []applicationEndpoint {
+	nilEmpty := func(s *string) string {
+		if s == nil {
+			return ""
+		}
+		return *s
+	}
+	var endpoints []applicationEndpoint
+	err := s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.Query(`
+SELECT ae.charm_relation_uuid, s.name
+FROM application_endpoint ae
+LEFT JOIN space s ON s.uuid=ae.space_uuid
+WHERE ae.application_uuid=?`, s.constants.appID)
+		defer func() { _ = rows.Close() }()
+		if err != nil {
+			return errors.Capture(err)
+		}
+		for rows.Next() {
+			var uuid string
+			var name *string
+			if err := rows.Scan(&uuid, &name); err != nil {
+				return errors.Capture(err)
+			}
+			endpoints = append(endpoints, applicationEndpoint{
+				charmRelationUUID: uuid,
+				spaceName:         nilEmpty(name),
+			})
+		}
+		return nil
+	})
+	c.Assert(err, jc.ErrorIsNil, gc.Commentf("(Assert) Failed to fetch endpoints: %v", err))
+	return endpoints
+}
+
+func (s *applicationEndpointStateSuite) fetchApplicationExtraEndpoints(c *gc.C) []applicationEndpoint {
+	nilEmpty := func(s *string) string {
+		if s == nil {
+			return ""
+		}
+		return *s
+	}
+	var endpoints []applicationEndpoint
+	err := s.TxnRunner().StdTxn(context.Background(), func(ctx context.Context, tx *sql.Tx) error {
+		rows, err := tx.Query(`
+SELECT ae.charm_extra_binding_uuid, s.name
+FROM application_extra_endpoint ae
+LEFT JOIN space s ON s.uuid=ae.space_uuid
+WHERE ae.application_uuid=?`, s.constants.appID)
+		defer func() { _ = rows.Close() }()
+		if err != nil {
+			return errors.Capture(err)
+		}
+		for rows.Next() {
+			var uuid string
+			var name *string
+			if err := rows.Scan(&uuid, &name); err != nil {
+				return errors.Capture(err)
+			}
+			endpoints = append(endpoints, applicationEndpoint{
+				charmRelationUUID: uuid,
+				spaceName:         nilEmpty(name),
+			})
+		}
+		return nil
```

Comment:
As you're not returning the rows.Close() you should do the following:

```suggestion
		return rows.Err()
```